license.go

```diff
@@ -58,20 +58,66 @@ func licenseType(name string) Type {
 	return Other
 }
 
+// A phrase is a sequence of words used as a key for startIndexes.
+// Empirically, two words are best; more is slower.
+type phrase [2]int32
+
 type license struct {
 	typ          Type
 	name         string
+	text         string
 	doc          *document
-	startIndexes map[string][]int
+	startIndexes map[phrase][]int
 }
 
 type document struct {
-	text    []byte   // Original text.
-	words   []string // Normalized words.
-	byteOff []int32  // ith byteOff is byte offset of ith word in original text.
+	text    []byte  // Original text.
+	words   []int32 // Normalized words (indexes into c.words)
+	byteOff []int32 // ith byteOff is byte offset of ith word in original text.
+}
+
+// A Checker matches a set of known licenses.
+type Checker struct {
+	licenses []license
+	urls     map[string]string
+	dict     map[string]int32 // dict maps word to index in words
+	words    []string         // list of known words
+}
+
+// A License describes a single license that can be recognized.
+// At least one of the Text or the URL should be set.
+type License struct {
```

Comment:
Describe how the type is derived from Name, since it shows up in Match.Type.

Reply:
I would suggest just not using the type at all, honestly.

Reply:
We don't. But it's part of the API.

Reply:
I'll talk to Rob about it next time we meet. We can add a Type field later if we need to I suppose but first we need to figure out if the Type enum makes that much sense at all.